### PR TITLE
infrastructure proxy call for artifacts API

### PR DESCRIPTION
**Pull request by pnaik1**
Closes: https://issues.redhat.com/browse/RHOAIENG-9677

## Description
Added feature flag `disableArtifactsAPI`
Added proxy call
added hook to return different  object based on API call 

## How Has This Been Tested?
manually console logging the useArtifactStorage function

## Test Impact
added jest tests for hooks

## Request review criteria:

Self checklist (all need to be checked):
- [x] The developer has manually tested the changes and verified that the changes work
- [x] Commits have been squashed into descriptive, self-contained units of work (e.g. 'WIP' and 'Implements feedback' style messages have been removed)
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [ ] The developer has added tests or explained why testing cannot be added (unit or cypress tests for related changes)

If you have UI changes: 

- [ ] Included any necessary screenshots or gifs if it was a UI change.
- [ ] Included tags to the UX team if it was a UI/UX change (find relevant UX in the [SMEs](https://github.com/opendatahub-io/odh-dashboard/tree/main/docs/smes.md) section).

After the PR is posted & before it merges:
- [ ] The developer has tested their solution on a cluster by using the image produced by the PR to `main`

### CODE REVIEW

File: frontend/src/concepts/pipelines/apiHooks/useArtifactStorage.ts [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+  const [objectSize, setObjectSize] = useState<number | undefined>(undefined);
+  const artifactFeatureFlagEnabled = s3EndpointAvailable || artifactApiAvailable;
+  const [artifactResponse] = useGetArtifactById(artifactId, 'DOWNLOAD');
+  const artifactSize = Number(artifactResponse?.artifact_size);
+
+  useEffect(() => {
+    if (artifactApiAvailable) {
+      if (artifactResponse?.download_url) {
+        fetch(artifactResponse.download_url, {
+          headers: { 'Content-Type': 'text/plain' },
```
Comment (pnaik1) [inline frontend/src/concepts/pipelines/apiHooks/useArtifactStorage.ts:39]: @DaoDaoNoCode @Gkrumbach07 I am getting here `CORS` while fetching the download_url

File: frontend/src/concepts/pipelines/apiHooks/useArtifactStorage.ts [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+  const path = artifact?.getUri();
+  const artifactId = artifact?.getId().toString();
+
+  const [downloadedObject, setDownloadedObject] = useState<string | undefined>(undefined);
+  const [objectUrl, setStorageUrl] = useState<string | undefined>(undefined);
+  const [objectSize, setObjectSize] = useState<number | undefined>(undefined);
+  const artifactFeatureFlagEnabled = s3EndpointAvailable || artifactApiAvailable;
+  const [artifactResponse] = useGetArtifactById(artifactId, 'DOWNLOAD');
+  const artifactSize = Number(artifactResponse?.artifact_size);
```
Comment (pnaik1) [inline frontend/src/concepts/pipelines/apiHooks/useArtifactStorage.ts:33]: Reason to add this:  type incompatibility in `size` between fetchStorageObjectSize and `artifact_size`

File: frontend/src/concepts/pipelines/apiHooks/useArtifactStorage.ts [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+  const s3EndpointAvailable = useIsAreaAvailable(SupportedArea.S3_ENDPOINT).status;
+  const artifactApiAvailable = useIsAreaAvailable(SupportedArea.ARTIFACT_API).status;
+  const { namespace } = usePipelinesAPI();
+
+  const path = artifact?.getUri();
+  const artifactId = artifact?.getId().toString();
+
+  const [downloadedObject, setDownloadedObject] = useState<string | undefined>(undefined);
+  const [objectUrl, setStorageUrl] = useState<string | undefined>(undefined);
+  const [objectSize, setObjectSize] = useState<number | undefined>(undefined);
```
Comment (DaoDaoNoCode) [inline frontend/src/concepts/pipelines/apiHooks/useArtifactStorage.ts:30]: Could be simplified:
```suggestion
  const [downloadedObject, setDownloadedObject] = useState<string>();
  const [objectUrl, setStorageUrl] = useState<string>();
  const [objectSize, setObjectSize] = useState<number>();
```

File: frontend/src/concepts/pipelines/apiHooks/useArtifactStorage.ts (hunk 1 above)
Comment (Gkrumbach07) [reply frontend/src/concepts/pipelines/apiHooks/useArtifactStorage.ts:39]: I am not seeing that on my end. how are you testing it

Comment (pnaik1) [reply frontend/src/concepts/pipelines/apiHooks/useArtifactStorage.ts:39]: @Gkrumbach07 , let me correct my statement
while fetching `download_url`, I am getting `cors` error 

Screenshot 2024-07-22 at 6 02 32 PM

Comment (Gkrumbach07) [reply frontend/src/concepts/pipelines/apiHooks/useArtifactStorage.ts:39]: I did this and it worked

```tsx
const [artifactResponse, loaded, error] = useGetArtifactById(artifact.getId(), 'DOWNLOAD');
  console.log(artifactResponse, loaded, error);

  useEffect(() => {
    if (artifactResponse?.download_url) {
      fetch(artifactResponse.download_url, {
        headers: { 'Content-Type': 'text/plain' },
      })
        .then((response) => response.text())
        .then((downloadObject) => {
          console.log(downloadObject);
        })
        .catch((e) => console.log(e));
    }
  }, [artifactResponse, namespace]);
```

I can try again with the useArtifactStorage. or if you have a testing commit i can look at

**Review by DaoDaoNoCode — COMMENTED** [review]
Maybe we still need to make `downloadObject` a function as `fetchDownloadObject` or something else because I saw this https://github.com/red-hat-data-services/data-science-pipelines/blob/master/backend/api/v2beta1/swagger/artifacts.swagger.json#L102. The link will expire in 15 seconds, so we should fetch it every time when we are going to use it. Sorry for the misleading before.

File: frontend/src/concepts/pipelines/types.ts [hunk 4]
```diff
@@ -68,6 +96,11 @@ export type CreatePipelineRecurringRun = (
   data: CreatePipelineRecurringRunKFData,
 ) => Promise<PipelineRecurringRunKFv2>;
 export type GetExperiment = (opts: K8sAPIOptions, experimentId: string) => Promise<ExperimentKFv2>;
+export type GetArtifact = (
+  opts: K8sAPIOptions,
+  artifactId: string,
```
Comment (Gkrumbach07) [inline frontend/src/concepts/pipelines/types.ts:101]: can artifactId be a number here or did that have issues

File: frontend/src/concepts/pipelines/apiHooks/useArtifactStorage.ts [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+    }
+    const path = artifact.getUri();
+    const uriComponents = extractS3UriComponents(path);
+    if (uriComponents) {
+      return fetchStorageObjectSize(namespace, uriComponents.path);
+    }
+    return Promise.reject();
+  };
+
+  const getStorageObjectUrl = async (artifact: Artifact): Promise<string | undefined> => {
```
Comment (DaoDaoNoCode) [inline frontend/src/concepts/pipelines/apiHooks/useArtifactStorage.ts:72]: Do we also want to return either `artifactStorage.uri` or `artifact.getUri()` here, which will be used to show the text of the link?

Screenshot 2024-07-24 at 11 05 00 AM

Comment (Gkrumbach07) [reply frontend/src/concepts/pipelines/apiHooks/useArtifactStorage.ts:72]: I thought about that, but the user calling this hook needs to give the artifact anyway so they would have the uri. But your call

Comment (DaoDaoNoCode) [reply frontend/src/concepts/pipelines/apiHooks/useArtifactStorage.ts:72]: Yeah that makes sense, I am OK with not having the URI here.

File: frontend/src/concepts/pipelines/types.ts (hunk 4 above)
Comment (pnaik1) [reply frontend/src/concepts/pipelines/types.ts]: @Gkrumbach07 , I have changed it to `number`

**Comment by Gkrumbach07** [comment]
looks good

/lgtm

File: docs/dashboard-config.md
Comment (DaoDaoNoCode) [inline docs/dashboard-config.md:1]: Why are you making this change?

File: docs/dashboard-config.md [hunk 6]
```diff
@@ -9,7 +9,7 @@ By default the ODH Dashboard comes with a set of core features enabled that are
 The following are a list of features that are supported, along with there default settings.
 
 | Feature                      | Default | Description                                                                                          |
-|------------------------------|---------|------------------------------------------------------------------------------------------------------|
+| ---------------------------- | ------- | ---------------------------------------------------------------------------------------------------- |
```
Comment (DaoDaoNoCode) [inline docs/dashboard-config.md:12]: Is there a reason you made this change here?

**Review by DaoDaoNoCode — APPROVED** [review]
/lgtm
Just one question out of curiosity but I think that's not important.

**Pull request merged**